### ISSUE TITLE
Mpileup indel only output

### ISSUE THREAD

**Issue by apallav** [samtools/samtools#1585]
Hi, 

Is it possible to add “-output-indel” to enable making insertion,deletions calls exclusively?

Currently I am able to make use of “no-output-ins” or  “no-output-dels” to call base substitutions exclusively but not vice versa.

Appreciate your time and consideration,

Aparna

**Comment by apallav** [comment]
Thanks for getting back.

I am building an application that constructs umi families extracting umi tags from read ids given by  output-QNAME.  Currently I am reading through a coordinate for base substitutions , reference base and corresponding reads and  as you can imagine, previous coordinate for indel changes, corresponding reads to calculate allele frequencies with umi error correction.

Computing wise I am using perl’s emulation of Unix fork to accomplish parallel processing, but visiting 2 positions for calling variant at a coordinate is becoming cumbersome. 

It will be extremely useful for us if indel only feature is possible just like no indel option. I don’t  have  thorough understanding of Mpileup code, or leveraging htslib C library so only option for me is to use Mpileup capabilities, which I am extremely impressed with and have been using for my custom application for years now.

 >> To remove base substitutions and retain only indels means most entries for this column would be empty, which then means we need a special symbol for absent (eg .). Your parsing would need to take that into account too.

Yes, I understand. Currently looping over just to get indel changes and capture corresponding reads is costing computing wise and time consuming. 

Please advise.

**Comment by jkbonfield** [comment]
I'm afraid I still don't understand what it is you're doing.  Why does outputting indels only change the number of "visits" to the variant calling bit?  Maybe an example would help.

The bulk of the time spent by mpileup is in fetching the data and keeping pointers into the current position in all the sequences overlapping each reference coordinate.  The actual display of the bases is fairly low on performance criteria.

If you're finding the parsing of the results is slower than the pileup algorithm itself, then that would imply any changes we make is unlikely to resolve things and you'd be better looking at why the output parsing is so slow. (Eg due to using Perl.)

**Comment by apallav** [comment]
>>Why does outputting indels only change the number of "visits" to the variant calling bit? Maybe an example would help.

Correct. # of visits remain same, but # of bases and therefore QNAME data will be exclusive to indels, therefore less regexp search and parsing. I am able to make use of no deletion/insertions options now and would appreciate the other way around feature.

Speed wise i am fine, getting around 100 records in around 1 min, at the rate of 60k to 100k reads at any given position.  I agree C would be ideal for my application. perl and python are comparable.

I will pursue using htslib API If it’s not possible to provide this feature in Mpileup. Just let me know.

Thanks!

**Comment by jkbonfield** [comment]
I still don't understand exactly what it is you're wanting.   Please provide an example (as requested).

For example, would not displaying non-indels mean all *rows* that don't have at least one indel are removed?
Or would not display indels mean all *sequences* that don't contain an indel are removed?

If the latter, what happens if all sequences are removed as no indel exists?  Does that also imply the entire row is removed, or should there be "*" in the appropriate columns?

This is the first you've mentioned QNAME, but I'll not that `--no-output-ins` (for example) has no impact at all on the QNAME fields, depth, etc.  It sounds like you want something different here that isn't simply the opposite of `--no-output-ins` but is something completely different.  Essentially another format entirely.  Is there a reason why VCF isn't appropriate?

In summary, unless we can see a compelling case for change, and obviously a clear description of what that change is, then it won't likely get done.  Note that writing your own code around Htslib's pileup interface however isn't too hard to create your own custom tool.  See https://github.com/samtools/htslib/blob/develop/test/pileup.c as an example starting point, at around 5Kb of code.

**Comment by jkbonfield** [comment]
Closing this as no clear information on the change requested and no recent replies.
